### DISCUSSION TITLE
[Shipped] Improving Mobile Screen Performance with TTID / TTFD

### DISCUSSION THREAD

**Discussion by zoesyc** [getsentry/sentry#60127] [Launch Feedback]
It’s no secret that mobile application performance significantly impacts whether a customer engages with your product. In 2019, [*Think with Google*](https://www.thinkwithgoogle.com/marketing-strategies/app-and-mobile/mobile-site-speed-importance/#:~:text=In%20retail%2C%20we've%20seen,rates%20by%20up%20to%2020%25.&text=This%20presents%20any%20business%20that,customers%20go%20elsewhere%20%E2%80%94%20and%20fast) found that “a one-second delay in mobile load times can impact conversion rates by up to 20%.” 

So we’re releasing a new troubleshooting workflow focused on mobile-centric metrics - specifically TTID (Time to Initial Display) and TTFD (Time to Full Display) for screen load performance. To quote [Android docs](https://developer.android.com/topic/performance/vitals/launch-time):

> Both are equally important, as TTID lets the user know that the app is loading, and TTFD is when the app is actually usable. If either of these is too long, the user might exit your app before it even fully loads.

Starting today, you can find **Mobile** in the sidebar under **Performance**.

![Screenshot 2023-11-16 at 10 53 28 AM](https://github.com/getsentry/sentry/assets/38118634/ce9b28fa-36e7-4d5f-a41f-d46cc03a5e25)

<h2>What’s New</h2>

- **Release Comparison:** Easily compare the screen load performance across 2 individual releases
- **Device Comparison:** Quickly identify if a screen is underperforming for high, medium, or low mobile devices
- **Screen Overview:** Skim through the performance stats for a single screen and dive into samples of poor TTID, TTFD, or lengthy spans

<h2>Additional Requirements</h2>

- This release is only available to **Android** and **iOS** native applications
- **Minimum Sentry SDK versions:**
     - Android
         - 5.0 or higher for automatic activity transactions
         - 6.10.0 or higher to capture TTID
         - 6.17.0 or higher to capture TTFD (released March 2023)
    - Apple / iOS
         - 3.13 or higher to capture UIViewController transactions
         - 8.4.0 or higher to capture TTID and TTFD (released April 2023)
- **TTFD** requires you to manually wrap the screens you are interested in tracking. You can find more details on how to set this up in our docs ([Android](https://docs.sentry.io/platforms/android/performance/instrumentation/automatic-instrumentation/?original_referrer=https%3A%2F%2Fdocs.sentry.io%2Fproduct%2Fperformance%2Ffilters-display%2Fwidgets%2F#time-to-full-display), [Apple / iOS](https://docs.sentry.io/platforms/apple/performance/instrumentation/automatic-instrumentation/?original_referrer=https%3A%2F%2Fdocs.sentry.io%2Fproduct%2Fperformance%2Ffilters-display%2Fwidgets%2F#time-to-full-display))

To be able to use Mobile Screens, you'll need to have Performance set up - here's our [setup guide](https://docs.sentry.io/product/performance/getting-started/) for instructions. You can learn more about this release in our [docs](https://docs.sentry.io/product/performance/mobile-vitals/?original_referrer=https%3A%2F%2Fdocs.sentry.io%2Fproduct%2Fperformance%2Ffilters-display%2Fwidgets%2F#mobile).

<h2> We want to hear from you!</h2>
We're rethinking our mobile performance workflows and we'd like to know if this is working for you! We'd like to better understand:
<br></br>

1. What are your thoughts on a single consolidated performance score for Mobile?
(This would mean that Sentry would calculate a single score by weighing the normalized scores across industry recognized mobile metrics like start times, screen load times, app responsiveness. To get an idea of what this would look like, check out our recent launch for [Web Vitals](https://github.com/getsentry/sentry/discussions/59620))
2. What are your most urgent mobile performance concerns (e.g. App Launch Sequence, UI jank, Interactions, etc.)?
3. What are some additional features you'd like to see?

Feel free to share your thoughts in the discussion below - or you can find us on [Discord](https://discord.com/invite/Ww9hbqr) and [X](https://twitter.com/getsentry).

**Comment by marandaneto** [comment]
[flashlight](https://github.com/bamlab/flashlight) does something similar but it's a pre-release tool, it'd be cool if it's possible to do something similar but derived from production events (transactions/errors...) and filterable by device type, brand, etc.
That was one of my hack week ideas but didn't get traction with more people 😅

**Reply by zoesyc** [reply]
@marandaneto thanks for sharing! The concept of a unified performance for mobile is something we're exploring and from a limited number of early conversations there seems to be a lot of interest! 

From what I can tell, flashlight uses RAM, CPU, FPS, and runtime as key metrics. Do you think the same applies for apps in production? To put it differently: What metrics do you think would be most important to capture in a mobile perf score?
cc @shruthilayaj

**Reply by marandaneto** [reply]
It's expensive (I guess?) to track those on production (monitoring CPU/memory/fps etc at all times, for the data to make sense, it cannot be sampled most likely).
Consider all the metrics Sentry already has, e.g.
* Crash free rate
* ANR rate
* App start
* Slow rendering
* Failed HTTP requests rate (not yet? when?)
* Failed spans/transactions
* Probably slow functions based on a threshold when using profiling
* Degradation from the previous releases (eg compare crash free rate releases and the current one)

Each of those would probably have a different weight though.
Those are just ideas that come to my mind.

**Reply by marandaneto** [reply]
Maybe I misunderstood it as a "Mobile vitals score", if it's only about performance, only the "performance" features (app start, slow rendering, slow HTTP requests - use threshold, slow functions - use threshold - use profiling, CPU, memory, fps...)

**Comment by zoesyc** [comment]
We now have TTID / TTFD monitoring for React Native applications! 🚀 

You can learn more here: https://github.com/getsentry/sentry-react-native/releases/tag/5.20.0
